### ISSUE TITLE
Add baseConfig option

### ISSUE THREAD

**Issue by otakustay** [stylelint/stylelint#3253]
Currently stylelint provides `configOverrides` option to override config from projects, however in some case what we actually need is to set a base config which is allowed to be overriden by projects's `.stylelintrc`

This is similar to [eslint's baseConfig](https://eslint.org/docs/developer-guide/nodejs-api#cliengine) options, and is really useful when encapsulating styelint into a cli tool with recommended rules (`extends` is not a good solution since it require install and configure stylelint project by project)

**Comment by jeddy3** [comment]
@otakustay Thanks for the request. 

> `extends` is not a good solution since it require install and configure stylelint project by project

Are you able to expand on the use case e.g. show more details of how a `baseConfig` option would work differently than using `extends` and why you wouldn't want to install stylelint for each project. This will help us understand how best to incorporate this feature.

**Comment by jeddy3** [comment]
@otakustay Thanks for writing this up. It seems like a valid use case.

I assume it'd work in the same way as `configOverrides`, in that it'd respect things like `configBasedir` and would allow the use of `extends` and `plugins` within the `baseConfig` [configuration object](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#the-configuration-object). Only, in this instance, the properties within the `baseConfig` object would be overridden by any properties within a `.stylelintrc` file (and then again by any properties within `configOverrides`). 

@stylelint/core  Any objections to welcoming a PR for this feature?

**Comment by Arcanemagus** [comment]
Since this is using the Node.js API, why not do your configuration manipulation in `fe-lint`?

For example over in `linter-stylelint` there are several manipulations done to the configuration before `stylelint.lint` is actually called. [This bit of code](https://github.com/AtomLinter/linter-stylelint/blob/e167a13c210f299263f8dc75155aefc6431e45a8/lib/index.js#L144-L160) shows some similar working with the configuration, you would just need to modify the `options.config` object directly.

So to be clear, you would [grab the configuration](https://github.com/AtomLinter/linter-stylelint/blob/e167a13c210f299263f8dc75155aefc6431e45a8/lib/index.js#L127) for the file in question, then if one was found you merge the found config with your default one (with eg. `Object.assign`).

(I'm actually not sure what the point of `baseConfig` is in ESLint either, it isn't used in `linter-eslint`, I think it's a holdover from the pre-1.0 model myself.)

**Comment by otakustay** [comment]
There are several reasons for this:

1. The format of configuration. eslint allows both `.json` and `.js` as configuration
2. The place of configuration. eslint allows config files in root directory and any sub directories, each directory can have its own specific config overrides, configurations will be looked upon directory tree
3. Cache should be expired when any configuration changes

I think stylelint is supporting or will suport these features, and in this case config resolving and overriding will be a quite complex step and it's nearly impossible for `fe-lint` to deal with each sub directory instead of just pass `files` option

**Comment by Arcanemagus** [comment]
> I think stylelint is supporting or will suport these features, and in this case config resolving and overriding will be a quite complex step and it's nearly impossible for fe-lint to deal with each sub directory instead of just pass files option

That's exactly why the example I linked shows how to have `stylelint` determine the configuration for you, so you don't have to worry about recreating all that logic 😉.

**Comment by ntwb** [comment]
> 2. The place of configuration. eslint allows config files in root directory and any sub directories, each directory can have its own specific config overrides, configurations will be looked upon directory tree

stylelint already supports this, see the test [here](https://github.com/stylelint/stylelint/blob/master/lib/__tests__/standalone.test.js#L234-L279) and the test fixtures for this test [here](https://github.com/stylelint/stylelint/tree/master/lib/__tests__/fixtures/config-per-file/a)

**Comment by ntwb** [comment]
Adapted script:
```js
const lintConfig = require('stylelint-config-us');
const stylelint = require('stylelint');
const report = await stylelint.lint(lintConfig);
console.log(report.output);
```


Example published _shared config _`stylelint-config-us` package:
```
const lintConfig = {
    config: {
        rules: {
            'at-rule-semicolon-newline-after': 'always'
        }
    },
    files: [
        process.cwd() + '/src/**/*.css')
    ]
};
```

Example `.stylelintrc` for a project that needs to modify the base `stylelint-config-us`:

```json
{
    "extends": "stylelint-config-us",
    "rules": {
        "at-rule-semicolon-newline-after": "never"
    }
}
```

The key change here is that for any project that needs to modify the base config, rather than just adding any rule changes required they first e_xtend_ your _stylelint shared config_ which should then do what I believe your looking to do and the rule per the stylelint [extends](https://stylelint.io/user-guide/configuration/#extends) functionality as expected.

The other benefit here is that your `stylelint-config-us` package can also be updated without the need to release a new version of `fe-lint`, dependant upon the semver changes.
